java/food-packaging/src/main/java/org/acme/foodpackaging/domain/Job.java

```diff
@@ -45,7 +45,7 @@ public class Job {
     /**
      * Start is after cleanup.
      */
-    @CascadingUpdateShadowVariable(targetMethodName = "updateStartCleaningDateTime", sourceVariableNames = {"line", "previousJob"})
+    @CascadingUpdateShadowVariable(targetMethodName = "updateStartCleaningDateTime")
     private LocalDateTime startCleaningDateTime;
 
     @PiggybackShadowVariable(shadowVariableName = "startCleaningDateTime")
```

Comment:
Didn't we discuss that _nothing_ is allowed to depend on cascading shadows?
How do piggybacks work, if the shadows are not part of the var listener chain anymore, but instead are processed independently in their own loop?

Reply:
No other shadow variable can use the cascading variable as a source, except for `@PiggybackShadowVariable`.

In that case, the parsing method will identify it as a cascading variable and add it as an additional shadow variable pointing to the same target method. This is equivalent:

```
@CascadingUpdateShadowVariable(targetMethodName = "updateStartCleaningDateTime")
private LocalDateTime startCleaningDateTime;

@CascadingUpdateShadowVariable(targetMethodName = "updateStartCleaningDateTime")
private LocalDateTime startProductionDateTime;

@CascadingUpdateShadowVariable(targetMethodName = "updateStartCleaningDateTime")
private LocalDateTime endDateTime;
```
I am not opposed to removing it, but doing so would make it more consistent with the other listeners.

Reply:
I don't have an opinion on whether there should be piggybacks, or if it should be the cascades. 

If they're piggybacks, then it implies that there is an order. First the cascade is updated, then the piggyback. Is that true?

Reply:
But what does it mean? What's the guarantee that the user has?

If something is a piggy-back, it means - implies - that first the cascade is updated, and only after that, the piggy-backs are updated. If this is not what's implemented, then arguably it shouldn't be a piggy-back.

But even if all of those are cascades - what is the actual order of update? "determined by user logic" isn't an answer, that doesn't say anything.

Reply:
> But what does it mean? What's the guarantee that the user has?
>
> If something is a piggy-back, it means - implies - that first the cascade is updated, and only after that, the piggy-backs are updated. If this is not what's implemented, then arguably it shouldn't be a piggy-back. (And if it is implemented like that, then in what order are multiple piggybacks updated for the same single cascade var?)
>

In this case, the main objective was to avoid the boilerplate from the previous implementation. The piggyback refers to updating the shadow variable using the same target method from the source. So, all cascading shadow variables and the related piggyback variables will be updated utilizing the same target method in one iteration.

In that case, there is no precedence relationship; it only causes the variable to point to the same cascading target method.
 
> But even if all of those are cascades, and not piggy-backs - what is the actual order of update? "determined by user logic" isn't an answer, that doesn't say anything.

Are we talking about the after-events generated by the changes or the variable updates inside the target method?

Reply:
I see what you mean, but there is always a precedence relationship. The code doesn't run concurrently, so one variable will physically be updated before another. What I'm asking here is for this physical ordering of variables to be described.

Reply:
Hmm, you're right. The cascades and the piggy-backs will be updated by the same method, so the order is not controlled by us in any way.

But this kinda leads me to ask: why do we annotate them at all? What purpose does it serve? The method will update anything it wants to; and no annotation will change anything about it.

Reply:
The main reason is that we can check if any of those variables have changed their values, which will result in an after-event and CS retesting the related constraints.

Reply:
The CS update happens on entity level though. So once _any_ variable on an entity is modified, the constraints will be re-evaluted. Which is why I think the piggybacks for cascade are useless.